File: heapsort.cpp
// Heapsort and Min heap in C++

#include<iostream>
#include<vector>

using namespace std;

#define ll long long 

void heapify(int i, vector<ll> &v)
{
	if(!v.size())
		return;
	if(i>=(v.size()+1)/2)
		return;

	int k=v[2*i]<v[(2*i)+1]?(2*i):(2*i+1);

	if(v[i]>v[k]){
	swap(v[i], v[k]);
	heapify(k, v);
	}
}

void build_heap(vector<ll> &v)
{
	if(!v.size())
		return;
	int i=(v.size()+1)/2;
	while(i--)
		heapify(i, v);
}

void insert(ll n, vector<ll> &v)
{
	if(!v.size())
		return;

	v.push_back(n);
	int i=v.size()-1;
	
	while(i && v[i/2]>v[i])
	{
		swap(v[i], v[i/2]);
		i/=2;
	}
}

ll extract_min(vector<ll> &v)
{
	if(!v.size())
		return -1;

	ll min=v[0];
	v[0]=v[v.size()-1];
	v.pop_back();
	heapify(0, v);

	return min;
}

ll find_min(const vector<ll> v){
	return v.size()?v[0]:-1;
}

vector<ll> heap_sort(vector<ll> v)
{
	vector<ll> w;
	w.reserve(v.size());

	build_heap(v);
	while(!v.empty())
		w.push_back(extract_min(v));
	
	return w;
}

int main()
{
	//driver code

	vector<ll> v = {12, 11, 13, 5, 6, 7}; 
    
	for(auto i : v)
		cout<<i<<" ";
	v=heap_sort(v); 

    	cout<<"\nSorted array is \n"; 
  
	for(auto i : v)
		cout<<i<<" ";

	cout<<endl;
	return 0;
}

File: heapsort.py
# Min-heap implemented in python

def heapify(node, heap):
    if len(heap)==0 or 2*node+1>=len(heap) or node<0:
        return 
    # next = 2*node+1 if heap[2*node+1]<heap[2*node+2] else 2*node+2
    try:
        next = 2*node+1 + int(heap[2*node+1]>heap[2*node+2])
    except IndexError:
        next = 2*node+1
    if heap[node]>heap[next]:
        heap[node], heap[next] = heap[next], heap[node]
        heapify(next, heap)

def extractMin(heap):
    if len(heap)==0:
        return None
    min = heap[0]
    heap[0] = heap[-1]
    heap.pop()
    heapify(0, heap)
    return min

def insert(ele, heap):
    heap.append(ele)
    point = len(heap)-1
    while heap[(point-1)//2]>heap[point] and point>=0:
        heap[(point-1)//2], heap[point] = heap[point], heap[(point-1)//2]
        point = (point-1)//2

def buildHeap(heap):
    for node in range((len(heap)+1)//2, -1, -1):
        heapify(node, heap)

def heapSort(heap):
    sorted=[]
    buildHeap(heap)
    while len(heap):
        sorted.append(extractMin(heap))
    
    return sorted

def main():
    #driver code
    arr = [ 12, 11, 13, 5, 6, 7] 
    arr = heapSort(arr) 
    n = len(arr) 
    print ("Sorted array is") 
    for i in range(n): 
        print ("%d"%arr[i]),

if __name__ == "__main__":
    main()